Sources/AsyncAlgorithms/AsyncMerge2Sequence.swift

```diff
@@ -9,201 +9,1065 @@
-      return try await apply(task1, task2)
-    case (.pending(let task1), .terminal):
-      return try await apply(task1, nil)
-    case (.terminal, .terminal):
-      return nil
+}
+
+extension AsyncMerge2Sequence: AsyncSequence {
+    public func makeAsyncIterator() -> AsyncIterator {
+        AsyncIterator(storage: internalClass.storage)
     }
-  }
 }
 
-extension Merge2StateMachine.Either where Base1.Element == Base2.Element {
-  var value : Base1.Element {
-    switch self {
-      case .first(let val):
-        return val
-      case .second(let val):
-        return val
+public extension AsyncMerge2Sequence {
+    struct AsyncIterator: AsyncIteratorProtocol {
+        /// This class is needed to hook the deinit to observe once all references to the ``AsyncIterator`` are dropped.
+        ///
+        /// If we get move-only types we should be able to drop this class and use the `deinit` of the ``AsyncIterator`` struct itself.
+        ///
+        /// - Important: This is safe to be unchecked ``Sendable`` since the `storage` is ``Sendable`` and `immutable`.
+        final class InternalClass: @unchecked Sendable {
+            private let storage: Storage
+
+            fileprivate init(storage: Storage) {
+                self.storage = storage
+                self.storage.iteratorInitialized()
+            }
+
+            deinit {
+                self.storage.iteratorDeinitialized()
+            }
+
+            func next() async throws -> Element? {
+                try await storage.next()
+            }
+        }
+
+        let internalClass: InternalClass
+
+        fileprivate init(storage: Storage) {
+            internalClass = InternalClass(storage: storage)
+        }
+
+        public mutating func next() async throws -> Element? {
+            try await internalClass.next()
+        }
     }
-  }
 }
 
-/// An asynchronous sequence of elements from two underlying asynchronous sequences
-///
-/// In a `AsyncMerge2Sequence` instance, the *i*th element is the *i*th element
-/// resolved in sequential order out of the two underlying asynchronous sequences.
-/// Use the `merge(_:_:)` function to create an `AsyncMerge2Sequence`.
-public struct AsyncMerge2Sequence<Base1: AsyncSequence, Base2: AsyncSequence>: AsyncSequence, Sendable
-where
-  Base1.Element == Base2.Element,
-  Base1: Sendable, Base2: Sendable,
-  Base1.Element: Sendable,
-  Base1.AsyncIterator: Sendable, Base2.AsyncIterator: Sendable {
-  public typealias Element = Base1.Element
-  /// An iterator for `AsyncMerge2Sequence`
-  public struct Iterator: AsyncIteratorProtocol, Sendable {
-    var state: Merge2StateMachine<Base1, Base2>
-    init(_ base1: Base1.AsyncIterator, _ base2: Base2.AsyncIterator) {
-      state = Merge2StateMachine(base1, base2)
+private extension AsyncMerge2Sequence {
+    final class Storage: @unchecked Sendable {
+        /// The lock that protects our state.
+        private let lock = Lock.allocate()
```

Comment:
I tried to adopt this but `ManagedCriticalState` does only support using a lock with a closure. For the `next` method I would like to manually lock and unlock to be able to hold the lock across the `withUnsafeContinuation` method.

Reply:
Holding a lock across a suspension is almost always a really bad idea; it can cause some pretty heinous fall out.

Reply:
Holding a lock across `with*Continuation` is safe to do since the closure is executed right away. We have been doing this in various places of the NIO ecosystem now.

Reply:
To further clarify this is not holding a lock across a suspension point but just across the `with*Continuation` call.

Reply:
so there is only 1 way to do that safely (and it is a very edge case scenario).

```
lock()
await withUnsafeContinuation { continuation in
  self.continuation = continuation
  unlock()
}
```

It is very easy to modify that pattern into some really gnarly failure modes. From a maintainability standpoint id rather make sure there is no other way of doing that.

Reply:
Right I am using exactly that pattern just with calls to the state inbetween. I agree with you that this needs to be done with care.
However, if don’t hold the lock across that call, the state handling becomes trickier since it introduces a potential interleaving between next and closure of withUnsafeContinuation where everything else can happen